packages/react-form/src/hooks/list/test/dynamiclist.test.tsx

```diff
@@ -0,0 +1,105 @@
+/* eslint-disable react/no-array-index-key */
+import React from 'react';
+import {mount} from '@shopify/react-testing';
+
+import {useDynamicList} from '../dynamiclist';
+import {FieldListConfig} from '../baselist';
+
+import {Variant, randomVariants, clickEvent, TextField} from './utils';
+
+const factory = () => {
+  return {price: '', optionName: '', optionValue: ''};
+};
+
+describe('useDynamicList', () => {
+  describe('add and remove fields', () => {
+    function DynamicListComponent(config: FieldListConfig<Variant>) {
+      const {fields, addItem, removeItem} = useDynamicList<Variant>(
+        config,
+        factory,
+      );
+
+      return (
+        <ul>
+          {fields.map((fields, index) => (
+            <li key={index}>
+              <TextField
+                label="price"
+                name={`price${index}`}
+                {...fields.price}
+              />
+              <button type="button" onClick={() => removeItem(index)}>
+                Remove Variant
+              </button>
+            </li>
+          ))}
+          <button type="button" onClick={() => addItem()}>
+            Add Variant
+          </button>
+        </ul>
+      );
+    }
+
+    it('can remove field', () => {
```

Comment:
To make sure nothing breaks it might worth adding another test case which handles the remove when the list is already empty.